src/codewhisperer/util/closingBracketUtil.ts

```diff
@@ -4,71 +4,123 @@
-                        editBuilder.delete(range)
-                    },
-                    { undoStopAfter: false, undoStopBefore: false }
-                )
-            }
+    }
+    return bracketsIndexLevel
+}
+
+export const getBracketsToRemove = (recommendation: string, rightContext: string) => {
+    const recommendationBrackets = calculateBracketsLevel(recommendation)
+    const rightContextBrackets = calculateBracketsLevel(rightContext)
+    let i = 0
+    let j = 0
+    const toBeRemoved = []
+
+    while (i < recommendationBrackets.length && j < rightContextBrackets.length) {
+        const { char: char1, count: level1 } = recommendationBrackets[i]
+        const { char: char2, count: level2, idx: idx2 } = rightContextBrackets[j]
+        if (char1 !== char2 || level1 !== level2) {
+            i++
+            continue
         }
+        toBeRemoved.push(idx2)
+        i++
+        j++
     }
+    return toBeRemoved
 }
 
-export function hasExtraClosingBracket(
-    recommendation: string,
-    openingBracket: string,
-    closingBracket: string
-): boolean {
-    let count = 0
-    let pos = recommendation.indexOf(closingBracket)
-    while (pos > 0) {
-        count++
-        pos = recommendation.indexOf(closingBracket, pos + 1)
+export const removeBracketsFromRightContext = async (
+    editor: vscode.TextEditor,
+    idxToRemove: number[],
+    endPosition: vscode.Position
+) => {
+    const offset = editor.document.offsetAt(endPosition)
+
+    if (isCloud9()) {
+        const edits = idxToRemove.map(idx => ({
+            range: new vscode.Range(
+                editor.document.positionAt(offset + idx),
+                editor.document.positionAt(offset + idx + 1)
+            ),
+            newText: '',
+        }))
+        const wEdit = new WorkspaceEdit()
+        wEdit.set(editor.document.uri, [...edits])
+        await workspace.applyEdit(wEdit)
+    } else {
+        await editor.edit(
+            editBuilder => {
+                idxToRemove.forEach(idx => {
+                    const range = new vscode.Range(
+                        editor.document.positionAt(offset + idx),
+                        editor.document.positionAt(offset + idx + 1)
+                    )
+                    editBuilder.delete(range)
+                })
+            },
+            { undoStopAfter: false, undoStopBefore: false }
+        )
     }
-    pos = recommendation.indexOf(openingBracket)
-    while (pos > 0) {
-        count--
-        pos = recommendation.indexOf(openingBracket, pos + 1)
+}
+
+export async function handleExtraBrackets(
+    editor: vscode.TextEditor,
+    recommendation: string,
+    endPosition: vscode.Position
+) {
+    const end = editor.document.offsetAt(endPosition)
+    const rightContext = editor.document.getText(
+        new vscode.Range(
+            editor.document.positionAt(end),
+            editor.document.positionAt(end + CodeWhispererConstants.charactersLimit)
+        )
+    )
```

Comment:
for getting the right context here, only the right context with a mixed string of consecutive brackets and empty characters(space, \t, \n) are needed. Once we detect a non-bracket or a non-empty character we can stop.

Reply:
To this end I don't think we have suppor to match whitespaces in this PR. It doesn't hurt to get that support. And it will be simpler than bracket matching, there will be no `bracketCounts` logic involved, whenever you find a space you just add it to the returned list.

Reply:
@andrewyuq what do you mean by matching whitespaces? could you pls provide an example? If it's formatting we do format the doc after accepting the recommendation

Reply:
yea, so let's say the right context is `) {`. If we just match brackets, the algorithm will stop at the first closing paramthese, if we can also match spaces it will match the the whole right context.

Reply:
@andrewyuq got it, yes the revised version does that, as you said we only stop if we encounter a non-bracket and non-white space character

Reply:
I get your point, I mean that we also need to **match** the whitespaces, let's say the recommendaiton is `){` and right context is `) {`. We can only match to the first `)` but not further, this means we also need to add the tripet entry for the whitespace character in `bracketsIndexLevel`(and probably should rename it as symbolsIndexLevel since it no longer contains brackets only). If we don't do this, the space will be ignored and they will be matched to the end by mistake.

Reply:
@andrewyuq I don't think we need to do that. if the recommendation is `){` and right context `) {` we will match them correctly. please see the description I will add a demo video for this case.

Reply:
>if the recommendation is `){` and right context `) {`
The thing is we actually don't want the recommendation to match the right context in this case. Notice that the recommendation does not have a whitespace and in the right context there's a whitespace. So the expected behavior should be that the matching should end after the first character `)`.

Reply:
Discussed offline, 1) The extra whitespace matching could cover more cases 2) whitespaces changing in Java and Javascript would change the syntatically correctness of the code, and in vscode we always do a pretty-print sort of formatting(vscode built-in) at the end, which should adapt user's local settings on the tab indent size in python. So there should be no problems matching the whitespaces in the insertion logic on acceptance.

Reply:
synced offline, we will match the whitespace in VSC because we are only modifying the right context after user accepting the recommendation and it will help to match more cases